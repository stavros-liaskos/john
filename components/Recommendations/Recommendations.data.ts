export const recommendationsI18n = {
  title: 'Other artists',
  artistList: {
    btnTxt: 'Unfollow',
    noArtists: 'You don not track any artists yet',
  },
};
