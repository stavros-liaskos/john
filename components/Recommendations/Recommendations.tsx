import ArtistsList, { ArtistsListI18n } from '../ArtistsList/ArtistsList';
import React, { useEffect, useState } from 'react';
import { components } from '../../types/schema';
import { followArtist } from '../../utils/followArtist';

type RecommendationsI18n = {
  title: string;
  artistList: ArtistsListI18n;
};

const Recommendations = ({ i18n }: { i18n: RecommendationsI18n }) => {
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const [recommendedArtists, setRecommendedArtists] = useState<components['schemas']['FollowedArtistDto'][]>([]);

  useEffect(() => {
    // TODO add Suspense for data fetching
    fetchRecommendations(setRecommendedArtists);
  }, []);

  if (!i18n || !i18n.title) {
    return null;
  }

  return (
    <div className="flex flex-col flex-2 basis-24 lg:justify-center items-center mb-2 w-full">
      <h3 className={'h3'}>{i18n.title}</h3>
      <ArtistsList
        i18n={i18n.artistList}
        artistsList={recommendedArtists}
        onButtonClick={handleFollow}
        artistLoading={artistLoading}
      />
    </div>
  );

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
    };
    await followArtist(artist, finallyCb);
  }
};

export function fetchRecommendations(
  setRecommendedArtists: React.Dispatch<React.SetStateAction<components['schemas']['FollowedArtistDto'][]>>,
) {
  fetch(`${process.env.BE_BASE_URL}/artists/recommended?page=0&size=10`, {
    method: 'GET',
    credentials: 'include',
  })
    .then(res => res.json())
    .then((result: components['schemas']['FollowedArtistsResponse']) => {
      result?.rows && setRecommendedArtists(result?.rows);
    })
    .catch(console.error);
}

export default Recommendations;
