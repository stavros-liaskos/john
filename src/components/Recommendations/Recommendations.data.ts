export const recommendationsI18n = {
  title: 'Others follow',
  artistList: {
    btnTxt: 'Follow',
    noArtists: 'There are no recommended artists for you :(',
  },
};
