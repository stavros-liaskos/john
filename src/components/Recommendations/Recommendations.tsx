import ArtistsList, { ArtistsListI18n } from '../ArtistsList/ArtistsList';
import React, { useState } from 'react';
import { components } from '../../types/schema';
import { followArtist } from '../../utils/followArtist';
import resources from '../../utils/Resources';

type RecommendationsI18n = {
  title: string;
  artistList: ArtistsListI18n;
};

const Recommendations = ({ i18n }: { i18n: RecommendationsI18n }) => {
  const [artistLoading, setArtistLoading] = useState<number>(0);
  const url = `${process.env.BE_BASE_URL}/artists/recommended?page=0&size=10`;

  const recommendedArtists = resources.fetch(url);

  if (!i18n || !i18n.title) {
    return null;
  }

  return (
    <div className="flex flex-col flex-2 basis-24 lg:justify-center items-center mb-2 w-full">
      <h3 className={'h3'}>{i18n.title}</h3>
      <ArtistsList
        i18n={i18n.artistList}
        artistsList={recommendedArtists.rows}
        onButtonClick={handleFollow}
        artistLoading={artistLoading}
      />
    </div>
  );

  async function handleFollow(artist: components['schemas']['SearchResultArtistDto']) {
    artist?.id && setArtistLoading(artist.id);

    const finallyCb = () => {
      artistLoading && setArtistLoading(0);
    };
    await followArtist(artist, finallyCb);
  }
};

Recommendations.whyDidYouRender = true;
export default Recommendations;
