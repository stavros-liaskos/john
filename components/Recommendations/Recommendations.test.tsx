import { render } from '@testing-library/react';
import Recommendations from './Recommendations';
import { recommendationsI18n } from './Recommendations.data';

describe('Recommendations', () => {
  it('renders without data without crashing', () => {
    // @ts-ignore
    render(<Recommendations />);
  });

  it('renders all elements', () => {
    const { getByText } = render(<Recommendations i18n={recommendationsI18n} />);

    expect(getByText(recommendationsI18n.title)).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { container } = render(<Recommendations i18n={recommendationsI18n} />);
    expect(container).toMatchSnapshot();
  });
});
