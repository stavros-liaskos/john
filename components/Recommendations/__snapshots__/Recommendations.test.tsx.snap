// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendations matches snapshot 1`] = `
<div>
  <div>
    <h3
      class="rr-text"
    >
      Other artists
    </h3>
    <div>
      artists lists todo
    </div>
  </div>
</div>
`;
